### ISSUE TITLE
`needless_borrow` suggestion doesn't compile

### ISSUE THREAD

**Issue by smoelius** [rust-lang/rust-clippy#8367]
### Summary

On the most recent nightly, the `needless_borrow` lint produces a suggestion that doesn't compile.

`cargo +nightly-2022-01-29 clippy` triggers this bug, but `cargo +nightly-2022-01-28 clippy` does not.

cc: @Jarcho (maybe?)

### Reproducer

**I tried this code:**

```rust
fn main() {
    let s0 = String::new();
    let mut chars = s0.chars();
    let s1 = (&mut chars).take(0).collect::<String>();
    println!("{}", s1);
    drop(chars);
}
```

**I expected to see this happen:**

Nothing.

**Instead, this happened:**

```
warning: this expression borrows a value the compiler would automatically borrow
 --> src/main.rs:4:14
  |
4 |     let s1 = (&mut chars).take(0).collect::<String>();
  |              ^^^^^^^^^^^^ help: change this to: `chars`
  |
  = note: `#[warn(clippy::needless_borrow)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_borrow
```
If you make the suggested change, you get:
```
error[E0382]: use of moved value: `chars`
    --> src/main.rs:6:10
     |
3    |     let mut chars = s0.chars();
     |         --------- move occurs because `chars` has type `Chars<'_>`, which does not implement the `Copy` trait
4    |     let s1 = chars.take(0).collect::<String>();
     |                    ------- `chars` moved due to this method call
5    |     println!("{}", s1);
6    |     drop(chars);
     |          ^^^^^ value used here after move
     |
note: this function takes ownership of the receiver `self`, which moves `chars`
    --> /home/smoelius/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/iter/traits/iterator.rs:1317:13
     |
1317 |     fn take(self, n: usize) -> Take<Self>
     |             ^^^^
```

### Version

```text
rustc 1.60.0-nightly (e0a55f449 2022-01-28)
binary: rustc
commit-hash: e0a55f4491a729bffc63b402ba903d90858c806b
commit-date: 2022-01-28
host: x86_64-unknown-linux-gnu
release: 1.60.0-nightly
LLVM version: 13.0.0
```

**Comment by ivan** [comment]
`needless_borrow` also started suggesting something that will compile but make the program segfault with a stack overflow :)

```rust
struct ThingA;
struct ThingB;

impl From<&ThingA> for ThingB {
    fn from(_: &ThingA) -> ThingB {
        ThingB
    }
}

impl From<ThingA> for ThingB {
    fn from(a: ThingA) -> ThingB {
        (&a).into()
    }
}

fn main() {
    let _: ThingB = ThingA.into();
}
```

```
warning: this expression borrows a value the compiler would automatically borrow
  --> src/main.rs:12:9
   |
12 |         (&a).into()
   |         ^^^^ help: change this to: `a`
   |
   = note: `#[warn(clippy::needless_borrow)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_borrow
```

**Comment by crumblingstatue** [comment]
Here is another false positive:
```Rust
use rand::distributions::Alphanumeric;
use rand::Rng;

#[derive(Debug)]
pub struct GameEntity {
    pub name: String,
    pub hp: i32,
    pub godmode: bool,
}

impl GameEntity {
    pub fn rand() -> Self {
        let mut rng = rand::thread_rng();
        let name_len = rng.gen_range(3..24);
        Self {
            name: (&mut rng)
                .sample_iter(&Alphanumeric)
                .take(name_len)
                .map(char::from)
                .collect(),
            hp: rng.gen_range(0..100),
            godmode: rng.gen(),
        }
    }
}
```
```
warning: this expression borrows a value the compiler would automatically borrow
  --> src/lib.rs:16:19
   |
16 |             name: (&mut rng)
   |                   ^^^^^^^^^^ help: change this to: `rng`
   ````

**Comment by Jarcho** [comment]
This will be fixed by #8355 whenever that gets merged.

**Comment by chris-morgan** [comment]
This issue is spawning quite a few duplicates: #8380, #8391, #8407, #8408, with multiple confirmations in several. #8355 (the patch that is supposed to fix this) doesn’t seem to be moving; it may need bumping, or else reverting whatever caused the regression, if possible.
